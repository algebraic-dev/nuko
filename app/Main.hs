module Main where

import GHC.IO.Encoding

import Syntax.Lexer.Support
import Syntax.Parser

import System.Environment

import qualified Data.ByteString as SB
import Data.Text.Encoding (decodeUtf8)
import Data.Text (pack, unpack)

import Syntax.Tree
import Error.Message
import Error.PrettyPrint
import Syntax.Expr

import Data.Set (toList)
import Type.Checker 
import Type.Types
import Type.Context


main :: IO ()
main = do
  setLocaleEncoding utf8
  [file] <- getArgs

  bs  <- SB.readFile file
  case runLexer parseExpr bs of
      Right res -> do
        putStrLn $ drawTree res
        print $ runGen $ do 
          (ty, ctx) <- exprTypeSynth [CtxAlpha "Int", CtxAlpha "String"] res
          pure (applyContext ctx ty)
      Left err -> putStrLn (unpack $ ppShow $ ErrReport (pack file) (decodeUtf8 bs) err)