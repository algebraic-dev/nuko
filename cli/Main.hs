module Main (main) where

import Relude
import Nuko.Driver            (compile, mainMod)
import Data.Aeson.Text        (encodeToLazyText)
import Data.Aeson             (ToJSON)
import Data.These             (These(These, That, This) )
import Nuko.Report.Render     (renderDiagnostic)

main :: MonadIO m => m ()
main = do
  content <- readFileBS "./examples/normal.nk"
  let res    = compile mainMod "examples/normal.nk" content
  let source = lines (decodeUtf8 content)
  case res of
    That _    -> putTextLn   "✓"
    These e _ -> putTextLn $ "•\n"  <> unlines (map (renderDiagnostic source) e)
    This e    -> putTextLn $ "✗\n"  <> unlines (map (renderDiagnostic source) e)