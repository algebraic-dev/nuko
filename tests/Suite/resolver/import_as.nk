type L =
  | A String Int
  | B Int String

import Main.L as E

let a (b: E.L) : String = 2