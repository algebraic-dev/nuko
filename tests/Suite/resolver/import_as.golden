✓ That
 └Prod
   ├Program
   | ├typeDecls:
   | | └List
   | |   └TypeDecl
   | |     ├tyName: (ReId L 0:5-0:6)
   | |     ├tyArgs: Empty List
   | |     └tyDecl:
   | |       └TypeSum
   | |         └NonEmpty
   | |           ├Prod
   | |           | ├ReId A 1:4-1:5
   | |           | └List
   | |           |   ├TId NoExt (Path "Prelude.String" 1:6-1:12)
   | |           |   └TId NoExt (Path "Prelude.Int" 1:13-1:16)
   | |           └Prod
   | |             ├ReId B 2:4-2:5
   | |             └List
   | |               ├TId NoExt (Path "Prelude.Int" 2:6-2:9)
   | |               └TId NoExt (Path "Prelude.String" 2:10-2:16)
   | ├letDecls:
   | | └List
   | |   └LetDecl
   | |     ├declName: (ReId a 6:4-6:5)
   | |     ├declArgs:
   | |     | └List
   | |     |   └Prod
   | |     |     ├ReId b 6:7-6:8
   | |     |     └TId NoExt (Path "Main.L.L" 6:10-6:13)
   | |     ├declBody: (Lit NoExt (LInt 6:26-6:27 2))
   | |     ├declRet:
   | |     | └Just (TId NoExt (Path "Prelude.String" 6:17-6:23))
   | |     └declExt: NoExt
   | ├impDecls: Empty List
   | └programExt: NoExt
   └LocalNS
     ├_openedNames:
     | └HashMap
     |   ├Prod
     |   | ├OccName TyName L
     |   | └Single Main.L L
     |   ├Prod
     |   | ├OccName TyName Int
     |   | └Single Prelude Int
     |   └Prod
     |     ├OccName TyName String
     |     └Single Prelude String
     ├_localNames: (NonEmpty HashMap)
     ├_openedModules:
     | └HashMap
     |   └Prod
     |     ├E
     |     └NameSpace
     |       ├_modName: Main.L
     |       └_names:
     |         └HashMap
     |           ├Prod
     |           | ├OccName ConsName A
     |           | └Public
     |           ├Prod
     |           | ├OccName TyName L
     |           | └Public
     |           └Prod
     |             ├OccName ConsName B
     |             └Public
     ├_currentNamespace:
     | └NameSpace
     |   ├_modName: Main
     |   └_names:
     |     └HashMap
     |       └Prod
     |         ├OccName VarName a
     |         └Public
     └_newNamespaces:
       └HashMap
         └Prod
           ├Main.L
           └NameSpace
             ├_modName: Main.L
             └_names:
               └HashMap
                 ├Prod
                 | ├OccName ConsName A
                 | └Public
                 ├Prod
                 | ├OccName TyName L
                 | └Public
                 └Prod
                   ├OccName ConsName B
                   └Public
